### PR TITLE
Don't recycle connections on RELOAD if TLS config is unchanged

### PR DESCRIPTION
**Pull request by cosgroveb**
This addresses #1145.

We run pgbouncers client-side that connect to 80+ databases. Separately we are addressing this bottleneck by running more pgbouncers client-side and striping them but in order to eliminate the unnecessary queuing that this kind of aggressive connection recycling can cause I propose taking a more judicious approach.

Essentially, I propose:

1. If going from TLS disabled => enabled, recycle.
2. If changing sslmode, recycle.
3. If changing TLS config, recycle.
4. Otherwise do nothing.

It's kind of gross that I have to inspect each member of the lengthy TLS config struct. Comments in-line.

### CODE REVIEW

File: src/sbuf.c
Comment (cosgroveb) [inline src/sbuf.c:32]: The fact that I need to include a header from libusual with "internal" in the name is a smell and a sign that maybe the logic to test whether a new TLS config structure matches an existing struct should actually live in there.

File: src/sbuf.c
Comment (cosgroveb) [inline src/sbuf.c:1241]: I wish C had reflection :) Alas. In any case, there are two members/structs I didn't check here worth noting. [`tls_error`](https://github.com/libusual/libusual/blob/master/usual/tls/tls_internal.h#L50) which seems to be for internal handling of TLS setup errors. Also, [`tls_keypair`](tls_keypair). keypair appears to be a list of keys. 

Maybe it should be checked? I would need to iterate through both the new and the old which is fine.

**Comment by cosgroveb** [comment]
3 out of 4 of my new tests fail in CI under a `--without-openssl`build :| 

```FAILED test/test_operations.py::test_servers_disconnect_when_changing_sslmode - AssertionError
FAILED test/test_operations.py::test_servers_disconnect_when_changing_tls_config - AssertionError
FAILED test/test_operations.py::test_servers_disconnect_when_enabling_ssl - AssertionError
======= 3 failed, 144 passed, 9 skipped, 9 warnings in 96.52s (0:01:36) ========
```

Maybe I'm missing a marker/annotation. 👀

**Comment by cosgroveb** [comment]
Okay, moved my tests into `test_ssl` so they only run in the right builds and addressed the uncrustify build.

File: src/sbuf.c
Comment (jcoleman) [inline src/sbuf.c:1239]: Is the struct sufficiently stable or only have binary comparable items in it such that we could simply mem compare? Or do we need to list out all of the attributes?

File: test/test_ssl.py
Comment (jcoleman) [inline test/test_ssl.py:386]: Maybe we could test that the list of backend connections itself (maybe measured from PG itself?) remains constant? The log message test to show it's 0 makes me a bit nervous about false green tests.

Comment (cosgroveb) [reply test/test_ssl.py:386]: That's a good idea. We can assert on both. The log assertions seemed to be conventional in this suite so I think there's some confidence in the approach but I like also asserting on connections from PG.

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c]: Cleaned up this syntax in [0adafa8](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/0adafa86c808be2e7c1dcdb485a65b5e7c2a0679)

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1239]: Interesting thought... I suspect that it's stable enough for that kind of comparison. I'll try it.

File: test/test_ssl.py
Comment (cosgroveb) [reply test/test_ssl.py:387]: Added [assertions](
f14b97d) to address this feedback

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1239]: aha yeah @jcoleman and I chatted and this doesn't work with a couple pointers to other structs and strings in `tls_config`

File: src/sbuf.c
Comment (JelteF) [reply src/sbuf.c:32]: Yeah, I think it's fine for initial review, but before merging the logic to compare a `tls_config` should be created as a PR of `libusual` instead (which can then be used in this PR).

File: src/sbuf.c
Comment (JelteF) [inline src/sbuf.c:1260]: nit, let's flip this around to avoid a double negation in the actual comparison and call it something `tls_changes_requires_reconnect`

File: src/sbuf.c
Comment (JelteF) [inline src/sbuf.c:1242]: There's a nasty problem here with this comparison. It seems likely that for rotations people might change the contents cert/key while keeping the path the same. This comparison wouldn't detect that. I don't know if we can actually compare this (especially when considering race conditions). So I think this limitation is probably fine, but we should document this in the user facing documentation. So for the `server_tls_ca_file` and `client_tls_ca_file` docs add something like this:
> If the contents of this file are changed without changing the actual setting, the new CA file will be used for new connections. Existing connections won't be closed though. If it's necessary for security reasons that connections using the old CA are recycled ASAP, it's advised to run `RECONNECT` after the reload.

File: test/test_ssl.py
Comment (JelteF) [inline test/test_ssl.py:384]: This should be in the next `log_contains` block (same for other tests)

File: test/test_ssl.py
Comment (JelteF) [reply test/test_ssl.py:409]: It indeed seems unnecessary to check this for each of these tests. But yeah let's keep one of them.

**Review by JelteF — CHANGES_REQUESTED** [review]
Overall looks like a good improvement. Currently we indeed recycle connections way too aggressively when TLS is enabled.

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1248]: Ah yes, it was on purpose but not necessarily for "good" reasons. See https://github.com/pgbouncer/pgbouncer/pull/1157#discussion_r1741782953

It can be worked through perhaps in the libusual PR.

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1260]: I like it thanks for the feedback on the double negation

File: test/test_ssl.py
Comment (cosgroveb) [reply test/test_ssl.py:409]: 0143b3d

File: test/test_ssl.py
Comment (cosgroveb) [reply test/test_ssl.py]: [fb9bed3](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/fb9bed316a0cd7b5605ef849df2654c11034bada). 

In the other tests there are `RELOAD`s in the `log_contains`block. The second and third test have one before the `log_contains`but only as part of test setup and the assertion block also has a `RELOAD` for each of those.

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1242]: This only checks the `server_tls_ca_file`, right?

Comment (cosgroveb) [reply src/sbuf.c:1242]: Ahh yep both are impacted by less aggressive connection recycling on `RELOAD` 11f099b

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c]: thanks fa1c177

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c]: [44b913b](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/44b913b7131abfeb20c05c3e09d0f629d096eb4f)

File: test/test_ssl.py
Comment (cosgroveb) [reply test/test_ssl.py]: Oops. I misread my tests and was wrong. Fixed. [5fe717a](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/5fe717a3b1fadac6469b0bfe6d3283960591e858)

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c:1247]: @JelteF I added some code to do this and a test that fails on [another branch here](https://github.com/pgbouncer/pgbouncer/commit/3e3fca58df9b50281ab2307aa3016911dbbccdd2).

Breakpointing in gdb it would appear that these sub-structs are NULLed out/not hydrated by libusual on config load... on the branch linked above:

```
(gdb) break sbuf.c:1274
Breakpoint 3 at 0x55555557dda3: file src/sbuf.c, line 1274.
(gdb) c
Continuing.
2024-09-04 14:16:43.208 UTC [1003231] LOG S-0x55555566ab80: pTxnPool/bouncer@127.0.0.1:10201 new connection to server (from 127.0.0.1:37876)
2024-09-04 14:16:48.748 UTC [1003231] LOG RELOAD command issued

Breakpoint 3, tls_config_unchanged (new_server_connect_conf=0x55555567e930)
    at src/sbuf.c:1274
1274            return (strings_equal(new_server_connect_conf->ca_file, server_connect_conf->ca_file) &&
(gdb) print new_server_connect_conf[0].keypair[0]
$22 = {next = 0x0, cert_file = 0x0, cert_mem = 0x0, cert_len = 0, key_file = 0x0, key_mem = 0x0, key_len = 0}
(gdb) print server_connect_conf[0].keypair[0]
$23 = {next = 0x0, cert_file = 0x0, cert_mem = 0x0, cert_len = 0, key_file = 0x0, key_mem = 0x0, key_len = 0}
(gdb)```

It's possible that these only get values when libusual needs to read in those files and create these objects.

I'll add a code comment and at this point it's time to move the work over to libusual.

Comment (cosgroveb) [reply src/sbuf.c:1250]: [1c373c5](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/1c373c58937bb5a376fba736cb093ceacaea3e81)

File: src/sbuf.c
Comment (cosgroveb) [reply src/sbuf.c]: d6ff3a09a0ae832cdf1c643f39ee626517297059

PRed here https://github.com/libusual/libusual/pull/57

**Comment by cosgroveb** [comment]
Hey @JelteF I think I've addressed any outstanding feedback. I did force push the branch on Wednesday to rewrite my git ùser.email`. Sorry if that was bad form or made reviewing the final state of this harder.

I presume we need to wait for my outstanding [libusual PR](https://github.com/libusual/libusual/pull/57) to be reviewed/merged to get this pull merged? Any tips on getting the libusual PR reviewed? 

Thanks for being so responsive and helpful in getting this operational bottleneck in pgbouncer solved.

**Comment by JelteF** [comment]
> I presume we need to wait for my outstanding https://github.com/libusual/libusual/pull/57 to be reviewed/merged to get this pull merged? Any tips on getting the libusual PR reviewed?

All PgBouncer committers also have commit access on libusual (including me). So this should be fine. But yeah we should merge that one first.

File: doc/config.md
Comment (JelteF) [inline doc/config.md:803]: Let's move this paragraph to the top of the TLS settings section. Also let's change the wording to include RELOAD, because now it sounds like this will happen automatically even without one. Something like this:
```suggestion
If the contents of any of the cert or key files are changed without changing the actual
setting filename in the config, the new file contents will be used for new connections after a RELOAD. Existing
connections won't be closed though. If it's necessary for security
reasons that all connections start using the new files ASAP, it's
advised to run RECONNECT after the RELOAD.
```

Comment (JelteF) [reply doc/config.md:803]: Probably also good to note that changing any of the settings in the config file will trigger a RECONNECT automatically.

Comment (cosgroveb) [reply doc/config.md]: [5e27dd2](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/5e27dd239c6f3ef22cf0dd379ab9b3ccdbb65ef4)

**Comment by cosgroveb** [comment]
@JelteF I've [branched a cleanup from this](https://github.com/cosgroveb/pgbouncer/tree/move_string_checking_func_to_libusual) to remove `strings_equal()` and use the libusual library function as [discussed there ](https://github.com/libusual/libusual/pull/57#issuecomment-2336713415)and will PR that as soon as the libusual PR and this one are merged.

**Comment by JelteF** [comment]
CI is reporting a failure on one of the newly added tests, looks like it's flaky since only one of the linux builds is hitting the issue.

**Comment by cosgroveb** [comment]
I haven't had success with any specs at all with valgrind enabled locally today to repro but will give it another try tonight

**Comment by cosgroveb** [comment]
Okay @JelteF these specs are genuinely flaky due to timing issues between `full_maint_period` refreshes clients checking out a connection. This spec was one that didn't sleep for `full_maint_period` and didn't expect that style of refresh from the logs but got one:

`2024-09-08 18:04:35.379 UTC [12865] LOG S-0x53ccd50: pTxnPool/bouncer@127.0.0.1:10201 closing because: database configuration changed (age=0s)`

This isn't incorrect behavior. Updating the expectation to avoid this: [e27c774](https://github.com/pgbouncer/pgbouncer/pull/1157/commits/e27c774b07e74372a5b82a5d9c2a235159d9f1dd)

**Comment by cosgroveb** [comment]
Any further changes needed @JelteF ?

**Review by JelteF — APPROVED** [review]
LGTM now

**Pull request merged**
